### ISSUE TITLE
Can using assert in the main programe code?

### ISSUE THREAD

**Issue by rushairer** [stretchr/testify#1429]
Thank you. 

I have a question:
Can you use github. com/stretchr/testify/assert in the main program code,  not test code only?
For example, when my program determines whether a struct is empty, it directly uses 'assert.IsEmpty'

If possible, do I need to pay attention to any areas to avoid performance issues?

**Comment by brackendawson** [comment]
You _can_, but you shouldn't.

Generally, assertions in this module always use reflection, which is slower than not using reflection, if the task you are doing can be performed naturally then that is always preferable. You also have to work out how to provide a `TestingT` value that wont panic. And your code is going to be less self-documenting and harder to read as a result.

Take your example of checking if a struct is "empty", probably the most efficient way to do this is to compare your value to an empty literal. You can do this if all fields in the struct are comparable. This is about two orders of magnitude faster than using testify:
```go
package kata

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

type Comparable struct {
	Name string
	Age  int
}

var v = Comparable{"Bracken", 21}

func BenchmarkNatural(b *testing.B) {
	for n := 0; n < b.N; n++ {
		if v == (Comparable{}) {
			b.FailNow()
		}
	}
}

func BenchmarkTestify(b *testing.B) {
	for n := 0; n < b.N; n++ {
		if !assert.NotEmpty(nil, v) {
			b.FailNow()
		}
	}
}
```

```
% go test -bench=.
goos: darwin
goarch: arm64
pkg: github.com/brackendawson/kata
BenchmarkNatural-10     1000000000               0.6314 ns/op
BenchmarkTestify-10     21137443                57.36 ns/op
PASS
ok      github.com/brackendawson/kata   2.793s
```

If your struct is not comparable, then you will find that using reflection is about the same speed as testify, but a specialised comparison, which may use type parameters (generics), is still an order of magnitude faster:
```go
package kata

import (
	"reflect"
	"testing"

	"github.com/stretchr/testify/assert"
	"golang.org/x/exp/slices"
)

type Incomparable struct {
	Name string
	Age  int
	Pets []string
}

func (c Incomparable) Equal(d Incomparable) bool {
	if c.Name != d.Name || c.Age != d.Age || !slices.Equal(c.Pets, d.Pets) {
		return false
	}
	return true
}

var v = Incomparable{"Bracken", 21, []string{"Shep"}}

func BenchmarkGeneric(b *testing.B) {
	for n := 0; n < b.N; n++ {
		if reflect.DeepEqual(v, Incomparable{}) {
			b.FailNow()
		}
	}
}

func BenchmarkSpecialised(b *testing.B) {
	for n := 0; n < b.N; n++ {
		if (Incomparable{}).Equal(v) {
			b.FailNow()
		}
	}
}

func BenchmarkTestify(b *testing.B) {
	for n := 0; n < b.N; n++ {
		if !assert.NotEmpty(nil, v) {
			b.FailNow()
		}
	}
}
```

```
% go test -bench=.
goos: darwin
goarch: arm64
pkg: github.com/brackendawson/kata
BenchmarkGeneric-10             17263496                70.25 ns/op
BenchmarkSpecialised-10         507728580                2.355 ns/op
BenchmarkTestify-10             22251014                53.90 ns/op
PASS
ok      github.com/brackendawson/kata   5.668s
```

In summary, please do not use testify outside of your tests.

**Comment by rushairer** [comment]
Thank you very much
